crypto/initthread.c

```diff
@@ -117,6 +117,9 @@ init_get_thread_local(CRYPTO_THREAD_LOCAL *local, int alloc, int keep)
             CRYPTO_THREAD_unlock(gtr->lock);
 #endif
             if (!CRYPTO_THREAD_set_local(local, hands)) {
+#ifndef FIPS_MODE
+                sk_THREAD_EVENT_HANDLER_PTR_pop(gtr->skhands);
```

Comment:
The [push command](https://github.com/openssl/openssl/blob/13bc4e5fb7f5317691d4619f246ccdd811b314ed/crypto/initthread.c#L112) is guarded by a lock, but the pop command isn't. Is that correct?

Reply:
Fixed, I hope

Reply:
I wonder whether the code wouldn't be much more readable if you keep the FIPS and the NON-FIPS implementation completely separated, instead of sprinkling the code with these ifndefn's...

_(I hope I didn't make a mistake while ripping it apart. Please double check the code, if you intend to use it.)_
```C
static THREAD_EVENT_HANDLER **
init_get_thread_local(CRYPTO_THREAD_LOCAL *local, int alloc, int keep)
{
    THREAD_EVENT_HANDLER **hands = CRYPTO_THREAD_get_local(local);

    if (alloc) {
        if (hands == NULL) {
#ifndef FIPS_MODE
            GLOBAL_TEVENT_REGISTER *gtr;

            if ((hands = OPENSSL_zalloc(sizeof(*hands))) == NULL) {
                OPENSSL_free(hands);
                return NULL;
            }

            /*
             * The thread event handler is thread specific and is a linked
             * list of all handler functions that should be called for the
             * current thread. We also keep a global reference to that linked
             * list, so that we can deregister handlers if necessary before all
             * the threads are stopped.
             */
            gtr = get_global_tevent_register();
            if (gtr == NULL) {
                OPENSSL_free(hands);
                return NULL;
            }
            CRYPTO_THREAD_write_lock(gtr->lock);
            if (!sk_THREAD_EVENT_HANDLER_PTR_push(gtr->skhands, hands)) {
                OPENSSL_free(hands);
                CRYPTO_THREAD_unlock(gtr->lock);
                return NULL;
            }

            if (!CRYPTO_THREAD_set_local(local, hands)) {
                sk_THREAD_EVENT_HANDLER_PTR_pop(gtr->skhands);
                CRYPTO_THREAD_unlock(gtr->lock);
                OPENSSL_free(hands);
                return NULL;
            }

            CRYPTO_THREAD_unlock(gtr->lock);

#else /* FIPS_MODE */
            if ((hands = OPENSSL_zalloc(sizeof(*hands))) == NULL) {
                OPENSSL_free(hands);
                return NULL;
            }

            if (!CRYPTO_THREAD_set_local(local, hands)) {
                OPENSSL_free(hands);
                return NULL;
            }
#endif
        }
    } else if (!keep) {
        CRYPTO_THREAD_set_local(local, NULL);
    }

    return hands;
}
```

Reply:
BTW: the `OPENSSL_free(hands)` can be removed (was not part of your patch).

```
#else /* FIPS_MODE */
            if ((hands = OPENSSL_zalloc(sizeof(*hands))) == NULL) {
                OPENSSL_free(hands);
                return NULL;
            }
```

Reply:
I would rather *not* do that, that makes it hard to see what both modes have in common.

Reply:
However, I worked out something that have fewer cpp conditionals